File: backend/app/learning/quiz_blueprint.py
"""Deterministic quotas; batch size limits structured-output length and provider cost."""
TYPES = ('single', 'multiple', 'judge', 'fill', 'written')
BATCH_SIZE = 5


def default_counts(total):
    if total < 3:
        return {'single': total}
    minority = max(1, total // 5)
    return {'single': total - minority * 2, 'multiple': minority, 'judge': minority}


def batches(counts):
    remaining = dict(counts)
    result, current = [], {}
    while any(remaining.values()):
        for kind in TYPES:
            if remaining.get(kind, 0):
                current[kind] = current.get(kind, 0) + 1
                remaining[kind] -= 1
                if sum(current.values()) == BATCH_SIZE:
                    result.append(current)
                    current = {}
    if current:
        result.append(current)
    return result

File: frontend/src/services/quizBlueprint.ts
export const questionTypes = ['single', 'multiple', 'judge', 'fill', 'written'] as const
export type QuestionType = typeof questionTypes[number]
export type QuestionCounts = Record<QuestionType, number>
export const questionLabels: Record<QuestionType, string> = { single: '单选题', multiple: '多选题', judge: '判断题', fill: '填空题', written: '问答题' }
export const defaultCounts = (total = 5): QuestionCounts => {
  const minority = total < 3 ? 0 : Math.max(1, Math.floor(total / 5))
  return { single: total - minority * 2, multiple: minority, judge: minority, fill: 0, written: 0 }
}
export const countQuestions = (value: QuestionCounts) => questionTypes.reduce((sum, type) => sum + value[type], 0)
export const validCounts = (value: QuestionCounts) => questionTypes.every(type => Number.isInteger(value[type]) && value[type] >= 0 && value[type] <= 20) && countQuestions(value) >= 1 && countQuestions(value) <= 20
